src/Symfony/Component/Cache/Adapter/RedisAdapter.php

```diff
@@ -229,53 +229,70 @@ protected function doSave(array $values, $lifetime)
-                }
-                foreach ($redis as $h) {
-                    if (!$h->exec()) {
-                        $failed = false;
-                    }
-                }
-            } else {
-                $this->pipeline(function ($pipe) use ($serialized, $lifetime) {
-                    foreach ($serialized as $id => $value) {
-                        $pipe->setEx($id, $lifetime, $value);
-                    }
-                });
-            }
-        } elseif (!$this->redis->mSet($serialized)) {
-            return false;
+        if (0 >= $lifetime) {
+            $this->redis->mSet($serialized);
+
+            return $failed;
         }
 
+        $this->pipeline(function ($pipe) use (&$serialized, $lifetime) {
+            foreach ($serialized as $id => $value) {
+                $pipe('setEx', $id, array($lifetime, $value));
+            }
+        });
+
         return $failed;
     }
 
+    private function execute($command, $id, array $args, $redis = null)
+    {
+        array_unshift($args, $id);
+        call_user_func_array(array($redis ?: $this->redis, $command), $args);
+    }
+
     private function pipeline(\Closure $callback)
     {
-        if ($this->redis instanceof \Predis\Client) {
-            return $this->redis->pipeline($callback);
-        }
-        $pipe = $this->redis instanceof \Redis && $this->redis->multi(\Redis::PIPELINE);
+        $e = null;
+        $redis = $this->redis;
+
         try {
-            $e = null;
-            $callback($this->redis);
+            if ($redis instanceof \Predis\Client) {
+                $redis->pipeline(function ($pipe) use ($callback) {
+                    $this->redis = $pipe;
+                    $callback(array($this, 'execute'));
+                });
+            } elseif ($redis instanceof \RedisArray) {
+                $connections = array();
+                $callback(function ($command, $id, $args) use (&$connections) {
+                    if (!isset($connections[$h = $this->redis->_target($id)])) {
+                        $connections[$h] = $this->redis->_instance($h);
+                        $connections[$h]->multi(\Redis::PIPELINE);
+                    }
+                    $this->execute($command, $id, $args, $connections[$h]);
+                });
+                foreach ($connections as $c) {
+                    $c->exec();
+                }
+            } else {
+                $pipe = $redis->multi(\Redis::PIPELINE);
+                try {
+                    $callback(array($this, 'execute'));
+                } catch (\Exception $e) {
+                }
+                if ($pipe) {
+                    $redis->exec();
+                }
+                if (null !== $e) {
+                    throw $e;
+                }
+            }
         } catch (\Exception $e) {
         }
-        if ($pipe) {
-            $this->redis->exec();
-        }
+
+        $this->redis = $redis;
```

Comment:
this won't be reset properly in case ``Error`` is thrown instead, leaving the class in a broken state.

Instead of catching the exception, resetting and then throwing again, I suggest using ``try/finally`` instead for the resetting (we can do that in Symfony 3)